src/TicketContext.tsx

```diff
@@ -3,59 +3,207 @@ import React, {
+        isRefreshingTickets: false,
+      };
+    }
+    case 'ADD_RESERVATION': {
+      return {
+        ...prevState,
+        activeReservations: [
+          ...prevState.activeReservations,
+          action.reservation,
+        ],
+      };
+    }
+    case 'UPDATE_RESERVATIONS': {
+      return {
+        ...prevState,
+        activeReservations: action.activeReservations,
+      };
+    }
+  }
+};
+
+export type ActiveReservation = {
+  reservation: TicketReservation;
+  offers: ReserveOffer[];
+  paymentType: PaymentType;
+  paymentStatus?: PaymentStatus;
+};
+
+type TicketState = {
   refreshTickets: () => void;
-  activatePollingForNewTickets: () => void;
+  activatePollingForNewTickets: (reservation: ActiveReservation) => void;
+} & Pick<
+  TicketReducerState,
+  'activeReservations' | 'fareContracts' | 'isRefreshingTickets'
+>;
+
+const initialReducerState: TicketReducerState = {
+  fareContracts: [],
+  activeReservations: [],
+  isRefreshingTickets: false,
 };
 
 const TicketContext = createContext<TicketState | undefined>(undefined);
 
 const TicketContextProvider: React.FC = ({children}) => {
-  const [poll, setPoll] = useState(false);
+  const [{activeReservations, ...state}, dispatch] = useReducer(
+    ticketReducer,
+    initialReducerState,
+  );
+
+  const updateReservations = useCallback(
+    (reservation: ActiveReservation) =>
+      dispatch({
+        type: 'ADD_RESERVATION',
+        reservation,
+      }),
+    [dispatch],
+  );
 
-  const getFareContracts = useCallback(async function () {
+  const getPaymentStatus = useCallback(async function (
+    paymentId: number,
+  ): Promise<PaymentStatus | undefined> {
     try {
-      const {fare_contracts} = await listFareContracts();
-      return fare_contracts;
+      const payment = await getPayment(paymentId);
+      return payment.status;
     } catch (err) {
       console.warn(err);
     }
-  }, []);
+  },
+  []);
+
+  const pollPaymentStatus = useCallback(
+    async function () {
+      const updatedReservations = await Promise.all(
+        activeReservations.map(async (res) => {
+          const paymentStatus = await getPaymentStatus(
+            res.reservation.payment_id,
+          );
+          return {
+            ...res,
+            paymentStatus,
+          };
+        }),
+      );
+
+      dispatch({
+        type: 'UPDATE_RESERVATIONS',
+        activeReservations: updatedReservations.filter(
+          (res) => !isHandledPaymentStatus(res.paymentStatus),
+        ),
+      });
+    },
+    [activeReservations, getPaymentStatus],
+  );
+
+  const updateFareContracts = useCallback(
+    async function () {
+      try {
+        dispatch({type: 'SET_IS_REFRESHING_FARE_CONTRACTS'});
+        const {fare_contracts: fareContracts} = await listFareContracts();
+        dispatch({type: 'UPDATE_FARE_CONTRACTS', fareContracts});
+      } catch (err) {
+        console.warn(err);
+      }
+    },
+    [dispatch],
+  );
+
+  useInterval(
```

Comment:
This should still poll even if inactive?

Reply:
Oops good catch, should be disabled if there are no active reservations, or if there are, and every one is CAPTURE'd